### PR TITLE
End-to-End Philly Message Parser Workflow

### LINKED ISSUE #2416: Save endpoint can now save to a DB using the extended schema
## What needs to be done

Upgrade our saving flow to save to the extended DB schema if so configured.

## Why it needs to be done

This ticket puts all the pieces together from previous tickets to allow saving to a DB using the extended schema. We have the extra data from the message parser, an extended schema for setting up the DB, and env vars that should let us tell the system what kind of world it lives in. Lets put it all together and get savin'!

## Acceptance Criteria

Given the env vars `DATABASE_SCHEMA='extended'`, `DATABASE_TYPE='sqlserver'` and using the orchestration config `philadelphia-ecr-viewer` :
- [ ] Write out the full eCR data json object returned from the parser into the extended schema tables. _**Important notes below!!!!!**_
- [ ] Tests/Test updates for new and updated functions

## Technical Notes
This is a write change only, we don't need to read the full extended schema

Extended Schema details in file: `containers/ecr-viewer/sql/extended.sql`

**_!!IMPORTANT!!_** Some of the SQL dialects we need to interact with can't handle `datetime` data types, so in the extended schema these have been split into two columns DATE and TIME. 

Example: An eCR's `encounter_start_date` and `encounter_start_time` are two separate columns in the DB despite coming from the same string in the eCR

### PR DESCRIPTION
**Pull request by JNygaard-Skylight**
# PULL REQUEST

## Summary
Enables end-to-end use of the Philly Orchestration config. Specifically saving the extended metadata to SQL Server.

## Related Issue
Fixes #2416 

## Acceptance Criteria
Given the env vars `DATABASE_SCHEMA='extended'`, `DATABASE_TYPE='sqlserver'` and using the orchestration config `philadelphia-ecr-viewer` :
- [ ] Write out the full eCR data json object returned from the parser into the extended schema tables. _**Important notes below!!!!!**_
- [ ] Tests/Test updates for new and updated functions

## Additional Information
You will need the following eCR-viewer environment variables set:
```
# Metadata Config
METADATA_DATABASE_TYPE=sqlserver
METADATA_DATABASE_SCHEMA=extended
SQL_SERVER_USER=sa
SQL_SERVER_PASSWORD=Password1!
SQL_SERVER_SERVER=sqlserver
```

Primarily you can test this by running the `process-message` endpoint:
1) In `containers/orchestatrion` Start orchestration and the relevant services with `docker compose up trigger-code-reference-service message-parser-service ingestion-service fhir-converter-service validation-service orchestration-service ecr-viewer postgres sqlserver --build`
2) You'll want to remove all FHIR bundles that were added to your save source 
2) Hit the `message-process` endpoint with this body:
```
{
  "message_type": "ecr",
  "data_type": "ecr",
  "config_file_name": "philadelphia-ecr-viewer.json",
  "message": "<EICR_DATA>",
  "rr_data": "<RR_DATA>"
}
```
3) Then confirm that the FHIR bundle has been saved whereever you saved it (the Philly orchestration config saves it to S3, but you should be able to change that to Postgres or Azure), and the metadata has been saved to SQL Server.

To test just the `save-fhir-data` you need to first get metadata. The extended metadata is quite large so you will not want to write it manually. You can run the message-parser seperately by:
1) Follow the instructions in the `README.md` in `containers/message-parser` to start the service.
2) Hit the `localhost:8080/parse-message` with this body:
```
{
  "message_format": "fhir",
  "parsing_schema_name": "philly_ecr.json",
  "message": "<FHIR_BUNDLE_JSON>"
}
```
3) This should return a big JSON full of the metadata

0) You will need the following  ENV variables set:
```
1) In `containers/ecr-viewer` Start the eCR with `npm run local-dev-sqlserver`.
2) Hit the `save-fhir-data` endpoint with this body:
```
{
    "fhirBundle": "<FHIR_BUNDLE_JSON>",
    "metadata":  "<METADATA_JSON>",
    "saveSource": "postgres" // Or wherever you want
}
```
3) Then confirm the FHIR bundle is saved in Postgres, and the metadata is in the SQL Server.

## Checklist

- [ ] If this code affects the other scrum team, have they been notified? (In Slack, as reviewers, etc.)

[//]: # (PR title: Remember to name your PR descriptively!)

### CODE REVIEW

File: containers/ecr-viewer/src/app/api/save-fhir-data/save-fhir-data-service.ts [hunk 1]
```diff
@@ -185,16 +184,264 @@ export const saveFhirData = async (
...
+            .input(
+              "test_result_interpretation",
+              sql.VarChar(255),
+              "I DON'T KNOW WHAT THIS IS SUPPOSED TO BE",
+            )
+            .input(
+              "test_result_interpretation_code",
+              sql.VarChar(50),
+              "I DON'T KNOW WHAT THIS IS SUPPOSED TO BE",
+            )
```
Comment (BobanL) [inline containers/ecr-viewer/src/app/api/save-fhir-data/save-fhir-data-service.ts:351]: Does someone have an answer to where this data would come from?

File: containers/ecr-viewer/src/app/api/save-fhir-data/save-fhir-data-service.ts [hunk 2]
```diff
@@ -185,16 +184,264 @@ export const saveFhirData = async (
...
+    },
+  });
+
+  console.log(metadata);
+
+  const transaction = new sql.Transaction(pool);
+
+  if (process.env.METADATA_DATABASE_SCHEMA == "extended") {
+    try {
+      await transaction.begin();
```
Comment (BobanL) [inline containers/ecr-viewer/src/app/api/save-fhir-data/save-fhir-data-service.ts:213]: Are we ending the transaction anywhere?

File: containers/ecr-viewer/src/app/api/save-fhir-data/save-fhir-data-service.ts [hunk 3]
```diff
@@ -185,16 +184,264 @@ export const saveFhirData = async (
...
+      await pool
+        .request()
+        .input("eICR_ID", sql.VarChar(200), metadata.eicr_id)
+        .input("eicr_set_id", sql.VarChar(255), metadata.eicr_set_id)
+        .input(
+          "fhir_reference_link",
+          sql.VarChar(255),
+          "I DON'T KNOW WHAT THIS IS SUPPOSED TO BE",
+        )
```
[the remark this answers is not in the document]
Comment (JNygaard-Skylight) [reply containers/ecr-viewer/src/app/api/save-fhir-data/save-fhir-data-service.ts:215]: `saveFhirData` does not return a link. Just the eCR ID. I'll create a ticket for this.

File: containers/ecr-viewer/src/app/api/save-fhir-data/save-fhir-data-service.ts (hunk 1 above)
Comment (JNygaard-Skylight) [reply containers/ecr-viewer/src/app/api/save-fhir-data/save-fhir-data-service.ts:344]: I'm going to remove this from the insert and create a ticket to address it.

File: containers/ecr-viewer/sql/entrypoint.sh [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+# Start SQL Server in the background
+/opt/mssql/bin/sqlservr &
+
+# Wait for SQL Server to start (you can fine-tune the sleep time if needed)
+sleep 120
```
Comment (BobanL) [inline containers/ecr-viewer/sql/entrypoint.sh:7]: Does wait need to be 2 minutes? It kind of throws me off waiting an additional 2 minutes when starting to wait for the schema to be loaded.

**Review by BobanL — APPROVED** [review]
Works perfectly locally! Thanks for all your hard work on this 😸

**Pull request merged**